quipucords/api/aggregate_report/model.py

```diff
@@ -0,0 +1,275 @@
+    """
+    Aggregate various totals from the facts related to the given report ID.
+
+    The logic here is kind of messy because we do not surface all the necessary
+    details into individual fields in the "normalized" SystemFingerprint.
+    To minimize the number of iterations and trips back to the database, we iterate
+    through all SystemFingerprint data and conditionally tally up various fields
+    and then we iterate through all InspectGroup data (raw facts) and conditionally
+    tally up various fields that only exist there. This iteration logic could be
+    simplified significantly if we normalized more data into SystemFingerprint or
+    some other even flatter and more normalized data structure that DOES NOT have
+    JSONField-like fields that require additional iteration. Until that exists,
+    the logic here will remain somewhat long and messy.
+
+    Much of this logic could be made more readable by breaking up each individual
+    field (or some groups of related fields) into their own "sum(map(lambda ...))"
+    one-liners, but that would mean iterating over the facts many more times, and
+    that would cause execution time to grow linearly with the number of hosts and
+    facts, and that may get too costly too quickly. The current "ugly" implementation
+    iterates through the facts exactly once.
+
+    TODO Try to break up some to this to reduce the noqa: C901,PLR0912,PLR0915.
+    """
+    aggregated = AggregateReport()
+
+    os_by_name_and_version = defaultdict(lambda: defaultdict(int))
+    system_creation_dates = []
+    vmware_vms_by_host = defaultdict(list)
+
+    report = Report.objects.get(pk=report_id)
+    fingerprints = report.deployment_report.system_fingerprints.all()
+    for fingerprint in fingerprints:
+        # Yes, we want *all* OS name/version combos regardless of `is_rhel`.
+        # This means output may include non-RHEL distros (e.g. Ubuntu, Debian).
+        # Note: Ansible sources are fingerprinted to store the Ansible controller
+        # version (e.g. "4.4.4") in SystemFingerprint.os_version, and that can
+        # results strange in output here like {"unknown": {"4.4.4": 1}}.
+        # TODO Does Ansible controller's version really belong there?
+        # See also: FingerprintTaskRunner._process_ansible_fact.
+        os_by_name_and_version[fingerprint.os_name or "unknown"][
+            fingerprint.os_version or "unknown"
+        ] += 1
+
+        # TODO Is there a better way to determine the underlying source type?
+        # Is there always at least 1 fingerprint.sources? are they always the same type?
+        # Should I be doing something more clever like iterating through the list
+        # and checking for any presence of each source type?
+        # Could there be multiple different source types in the same fingerprint?
+        # If I made a bad assumption here, that might invalidate everything else below.
+        source_type = fingerprint.sources[0].get("source_type")
```

Comment:
you shouldn't expect a single source type from a fingerprint. 
Remember deduplication?
A fingerprint could be generated by multiple source types, so, if we are to follow with this approach, we need to have support for multiple source types - otherwise, we are just choosing one "randomly".